src/sandstorm/sandstorm-ip-bridge.c++

```diff
@@ -0,0 +1,181 @@
+
+// This program is useful for including in Sandstorm application packages where
+// the application itself is a legacy HTTP web server that does not understand
+// how to speak the Cap'n Proto interface directly.  This program will start up
+// that server and then redirect incoming requests to it over standard HTTP on
+// the loopback network interface.
+
+#include "sandstorm/sandstorm-ip-bridge.h"
+
+#include <kj/debug.h>
+#include <netinet/ip.h>
+#include <linux/netfilter_ipv4.h>
+#include <unistd.h>
+
+namespace sandstorm {
+  namespace ipbridge {
+
+  class RefcountedAsyncIoStream: public kj::Refcounted {
+  public:
+    explicit RefcountedAsyncIoStream(kj::Own<kj::AsyncIoStream>&& stream)
+        : stream(kj::mv(stream)) {}
+    kj::Own<kj::AsyncIoStream> stream;
+  };
+
+  static const int BUFFER_SIZE = 8192;
+
+  // TCP handling
+  struct AcceptedConnection {
+    class Downstream final: public ByteStream::Server {
+    private:
+      kj::Own<RefcountedAsyncIoStream> connection;
+
+    public:
+      explicit Downstream(kj::Own<RefcountedAsyncIoStream> && connection) :
+        connection(kj::mv(connection)) {}
+
+      kj::Promise<void> write(WriteContext context) {
+        auto params = context.getParams();
+        return connection->stream->write(params.getData().begin(), params.getData().size());
+      }
+
+      kj::Promise<void> done(DoneContext context) {
+        connection->stream->shutdownWrite();
+        return kj::READY_NOW;
+      }
+    };
+
+  private:
+    kj::Own<RefcountedAsyncIoStream> connection;
+    capnp::byte buffer[BUFFER_SIZE];
+    TcpPort::Client port;
+    kj::Own<ByteStream::Client> upstream;
+
+  public:
+    explicit AcceptedConnection(kj::Own<kj::AsyncIoStream>&& connectionParam, TcpPort::Client port)
+        : connection(kj::refcounted<RefcountedAsyncIoStream>(kj::mv(connectionParam))), port(port) {
+      auto request = port.connectRequest();
+      request.setDownstream(kj::heap<Downstream>(kj::addRef(*connection)));
+      upstream = kj::heap<ByteStream::Client>(request.send().getUpstream());
+    }
+
+    kj::Promise<void> start() {
+      return messageLoop();
+    }
+
+    kj::Promise<void> messageLoop() {
+      return connection->stream->tryRead(buffer, 1, BUFFER_SIZE)
+            .then([this] (size_t size) -> kj::Promise<void> {
+        if (size < 1) {
+          return upstream->doneRequest().send().then([] (auto args) {});  // EOF
+        }
+
+        auto request = upstream->writeRequest();
+        request.setData(kj::ArrayPtr<capnp::byte>(buffer, size));
+        return request.send().then([this] (auto args) {
+          return messageLoop();
+        });
+      });
+    }
+  };
+
+  TcpPort::Client getTcpClient(kj::Own<kj::AsyncIoStream>& connection, HackSessionContext::Client& session) {
+    auto request = session.getIpNetworkRequest().send().getNetwork().getRemoteHostRequest();
+
+    struct sockaddr destaddr;
+    uint size = sizeof(destaddr);  // TODO(soon): make static?
```

Comment:
I would argue against making it static. If you do then you're assuming that `connection` is always the same object which seems unexpected from the function signature.